qiskit/pulse/builder.py

```diff
@@ -1832,89 +1854,165 @@ def snapshot(label: str, snapshot_type: str = "statevector"):
 
-        qc = circuit.QuantumCircuit(2)
-        qc.cx(0, 1)
-        qc_transpiled = transpile(qc, optimization_level=3)
-        sched = schedule(qc_transpiled, backend)
+            backend = FakeBogotaV2()
 
-        with pulse.build(backend) as pulse_prog:
-                pulse.call(sched)
-                pulse.call(qc)
+            with pulse.build() as x_sched:
+                pulse.play(pulse.Gaussian(160, 0.1, 40), pulse.DriveChannel(0))
 
-    This function can optionally take parameter dictionary with the parameterized target program.
+            with pulse.build() as pulse_prog:
+                pulse.call(x_sched)
 
-    .. code-block:: python
+            print(pulse_prog)
 
-        from qiskit import circuit, pulse
+        The actual program is stored in the reference table attached to the schedule.
 
-        amp = circuit.Parameter('amp')
+        .. jupyter-execute::
 
-        with pulse.build() as subroutine:
-            pulse.play(pulse.Gaussian(160, amp, 40), pulse.DriveChannel(0))
+            print(pulse_prog.references)
 
-        with pulse.build() as main_prog:
-            pulse.call(subroutine, amp=0.1)
-            pulse.call(subroutine, amp=0.3)
+        In addition, you can call a parameterized target program with parameter assignment.
 
-    If there is any parameter name collision, you can distinguish them by specifying
-    each parameter object as a python dictionary. Otherwise ``amp1`` and ``amp2`` will be
-    updated with the same value.
+        .. jupyter-execute::
 
-    .. code-block:: python
+            amp = circuit.Parameter("amp")
 
-        from qiskit import circuit, pulse
+            with pulse.build() as subroutine:
+                pulse.play(pulse.Gaussian(160, amp, 40), pulse.DriveChannel(0))
 
-        amp1 = circuit.Parameter('amp')
-        amp2 = circuit.Parameter('amp')
+            with pulse.build() as pulse_prog:
+                pulse.call(subroutine, amp=0.1)
+                pulse.call(subroutine, amp=0.3)
 
-        with pulse.build() as subroutine:
-            pulse.play(pulse.Gaussian(160, amp1, 40), pulse.DriveChannel(0))
-            pulse.play(pulse.Gaussian(160, amp2, 40), pulse.DriveChannel(1))
+            print(pulse_prog)
 
-        with pulse.build() as main_prog:
-            pulse.call(subroutine, value_dict={amp1: 0.1, amp2: 0.2})
+        If there is a name collision between parameters, you can distinguish them by specifying
+        each parameter object in a python dictionary. For example,
+
+        .. jupyter-execute::
+
+            amp1 = circuit.Parameter('amp')
+            amp2 = circuit.Parameter('amp')
+
+            with pulse.build() as subroutine:
+                pulse.play(pulse.Gaussian(160, amp1, 40), pulse.DriveChannel(0))
+                pulse.play(pulse.Gaussian(160, amp2, 40), pulse.DriveChannel(1))
+
+            with pulse.build() as pulse_prog:
+                pulse.call(subroutine, value_dict={amp1: 0.1, amp2: 0.3})
```

Comment:
Is there a reason to allow the user to write a subroutine with a naming clash between its parameters?  Seems a bit odd.

Reply:
I can't think of why you would set out to write a schedule like the example, but I could see it coming up in more complex code, like calling different functions to generate different parts of a schedule or using references to different schedules and then using `inline_subroutines` to merge them into one big schedule block.

Reply:
That's totally fine by me, if there's a reason.  I really don't know any of the context - I was basically just reviewing by "smell".

Reply:
Yeah basically this kind of problem occurs because the pulse module uses `Parameter` object designed for circuit as-is. Usually pulse parameter is scoped by pulse name (this allows user to safely have the same name in a schedule, i.e., sequence of a pulse), and thus UUID mechanism is not necessary. In pulse programming, it will never happen to have the same parameter name in the same pulse. So proper model should be `Parameter(scope, name)`, rather than `Parameter(name, uuid)`.